torchdynamo/variables/functions.py

```diff
@@ -140,11 +140,29 @@ def bind_args(self, parent, args, kwargs):
                 elif self.source:
                     from .builder import VariableBuilder
 
-                    source = AttrSource(
-                        GetItemSource(AttrSource(self.source, "__closure__"), idx),
-                        "cell_contents",
-                    )
-                    result[name] = VariableBuilder(parent, source)(cell.cell_contents)
+                    side_effects = parent.output.side_effects
+                    if cell in side_effects:
+                        out = side_effects[cell]
+                    else:
+                        closure_cell = GetItemSource(
+                            AttrSource(self.source, "__closure__"), idx
+                        )
+                        closure_cell_contents = AttrSource(
+                            closure_cell, "cell_contents"
+                        )
+
+                        # cells are written to with "cell_contents",
+                        # so the source should just be the closure_cell, not its contents
+                        out = side_effects.track_cell_existing(closure_cell, cell)
+                        side_effects.store_cell(
+                            out,
+                            VariableBuilder(parent, closure_cell_contents)(
+                                cell.cell_contents
+                            ),
+                        )
```

Comment:
Sorry, I know this is ancient history, but the new code here looks worse than the old code in one sense: we're unconditionally writing to the cell on function creation, triggering side effects, but in fact we may not actually need to write the closure cell if there aren't any variables that will get mutated in the function. I wonder if there's a way to avoid unconditionally store_cell here (context: I'm rewriting cond() operator to be bytecode based and for simplicity I want to assert that there are no side effects, but benign use of closures in our test cases are triggering side effects.)

Reply:
I think what I am going to try is analyzing the bytecode to see if there is a `STORE_DEREF` on the relevant freevar. If there isn't I can assume I don't need to allocate a cell.

Reply:
Sorry, back again from #137456.

> I wonder if there's a way to avoid unconditionally store_cell here

Is the purpose of this `side_effects.store_cell` call solely to allow us look up its value via `side_effects.load_cell` later, since that's how Dynamo models cell content at the moment?

It does feel a bit strange that we need to generate a write side-effect for an _existing_ cell, I don't see `track_cell_existing` used anywhere else either. Can we avoid this by putting some special treatment to `load_cell` for these "existing cells" from the outside world?

I think that might be less complexity than the restarting mechanism? The latter has generated some subtle bugs in the Dynamo stack, see #137510 and the referenced PRs in its commit message.